### PR TITLE
Remove memref type from ONNX Ops

### PR DESCRIPTION
**Pull request by chentong319**
In concept, ONNX Ops are defined on Tensor type and MemRef type is introduced when ONNX Ops are lowered to krnl dialect. However, in the process of lowering, some ONNX Ops are generated with the intermediate lowering result, which has the MemRef type. The previous solution is to allow AnyMemRef into the ONNX Op definition. This solution mixed the concept and implementation of ONNX Ops. We have to handle memref type in transformation of ONNX Ops.
This PR uses UnrealizedConversionCastOp to cast the memref back to tensor when an ONNX Op is created in lowering. The conversion is added into the builder utilities to avoid massive code changing.  As a result of this PR, the memref type is removed from ONNX Op definition.

Changes include:

1. gen_onnx_mlir.py: remove memref
2. Add the type conversion in builders: onnx/DialectBuilder and ONNXToKrnlCommon.cpp.
4. Modify lit test.

Issues

- The tensor type is simply inferred from memref in the current implementation, not from the original tensor type. This approach works only for our current implementation. For more generic solution, our builder has to be changed to include original type and type converter  for conversion.
- Manually defined ONNX Op has not been updated

Future work

- Clearly define the interface of builder at different stage. For example, onnx builder always converts all input to tensor type, while memref builder always convert input to memref.
- Integrate the builder with conversion. Use the  type converter, or addSourceMaterialization and addTargetMaterialization hook.

### CODE REVIEW

File: src/Dialect/ONNX/DialectBuilder.cpp [hunk 1]
```diff
@@ -50,29 +50,44 @@ Value OnnxBuilder::matmul(Type Y, Value A, Value B, bool useGemm) const {
...
+Value OnnxBuilder::totensor(Value input) const {
+  if (input.getType().isa<TensorType>())
+    return input;
+  assert(input.getType().isa<MemRefType>() &&
+         "expect RankedMemref type when not a TensorType");
+  auto aTy = input.getType().cast<ShapedType>();
+  auto aTensorTy = RankedTensorType::get(aTy.getShape(), aTy.getElementType());
+  return b.create<UnrealizedConversionCastOp>(loc, aTensorTy, input)
+      .getResult(0);
+}
```
Comment (tungld) [inline src/Dialect/ONNX/DialectBuilder.cpp:91]: It looks non-trivial to convert MemRefType to TensorType when MemRefType has non-trivial affine_map. I think we will use MemRefType with affine_map more often in near future. So we should make sure we have no problem with that.

Comment (chentong319) [reply src/Dialect/ONNX/DialectBuilder.cpp:91]: Tensor type should be inferred from the type of the original inputs (of Tensor type) of ONNX Op, not the transformed ones (Operands in the conversion). Inference tensor type from memref type will be avoided.

Comment (tungld) [reply src/Dialect/ONNX/DialectBuilder.cpp:91]: > Tensor type should be inferred from the type of the original inputs (of Tensor type) of ONNX Op, not the transformed ones (Operands in the conversion)

I see. Thanks!

**Review by tungld — APPROVED** [review]
LGTM. I agree on removing MemRef from ONNX dialect and am looking forward to other PRs to fix the listed issues.

File: src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp [hunk 2]
```diff
@@ -28,6 +28,7 @@ Value OnnxToKrnlBuilder::reshape(
 
   ShapedType inputType = input.getType().cast<ShapedType>();
   Type elementType = inputType.getElementType();
+  MultiDialectBuilder<OnnxBuilder> create(b, loc);
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp:31]: I like to use create.*.* to explicitly know which builder is being used.

File: src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp [hunk 3]
```diff
@@ -447,10 +464,16 @@ Value foldOrEmitONNXTransposeOp(ConversionPatternRewriter &rewriter,
                          .getResult();
     free(resBuffer);
     free(inputBuffer);
-    return constVal;
-  } else
-    return rewriter.create<ONNXTransposeOp>(loc, resultType, input, permAttr)
-        .getResult();
+    return create.onnx.toMemref(constVal);
+  } else {
+    MultiDialectBuilder<KrnlBuilder, MathBuilder, MemRefBuilder, OnnxBuilder>
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp]: fixed

File: src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp:31]: yes, I like it.

File: src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp]: tx

**Pull request merged**